### DISCUSSION TITLE
Welcome to Oxygen.jl Discussions!

### DISCUSSION THREAD

**Discussion by ndortega** [OxygenFramework/Oxygen.jl#32] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and be open-minded. Remember that this is a community we
  build together 💪.

They say teamwork makes the dream work, which means I need YOUR help to make this project into the absolute best version of itself.

**Comment by HiperMaximus** [comment]
I think a good roadmap for Oxygen would be feature-parity with FastAPI from Python, one of the better Frameworks I've used and in my experience probably the easiest to pickup because of it's documentation and automatic documentation generation of the APIs

**Reply by ndortega** [reply]
I agree, that would definitely be a good place to start. Just in case you missed it, I just released v1.0.5 which added auto-generated documentation for each registered endpoint. It's pretty basic at the moment, it just handles paths, parameters, and some response types. But its a good start

**Reply by HiperMaximus** [reply]
It would be good to have a simple tutorial, for beginners

**Reply by ndortega** [reply]
Sure, what kind of tutorial were you thinking of? Any specific set of topics you'd think people would find helpful?
Would something like blog post that walks you through how to build a simple weather app or a todo-list application be good?

**Reply by HiperMaximus** [reply]
Both of them are good ideas, ideally, if the road is going for something like FastAPI, then copying its tutorial would be pretty amazing. Not assuming much knowledge of Julia or web develpment in general would be good as it would attract a wider userbase+I think most Julia users aren't in the web-develpment side of things, so it would be helpful there too. Also adding a comparison with other frameworks might encourage adoption, e.g. speed, ergonomics, integration, goals, etc.

**Reply by ndortega** [reply]
I just took a look at their tutorial section and I agree. It has a nice blend of "here's how the web works" and here's how you can easily do these things with FastApi. For now,  I'm thinking of focusing on making that tutorial rather than building out some sample apps. (Feel free to contribute some demo apps if you want)

Thanks for the suggestion, that's a really good point. I've been adding a lot of features over the past month and I can see how beginners would be intimidated by a massive README.md

**Comment by jiachengzhang1** [comment]
Great package!

Critical features for me are authentication, session, database related, frontend template ([Mustache.jl](https://github.com/jverzani/Mustache.jl)) etc, they are quite standard for web servers nowadays. Some benchmark and performance measures are good to have too

Some ideas on auto-generated documentation. IMO to improve usability, capture the routing and model information when the routes are created, similar to how FastAPI generates the swagger. e.g.
```julia
mutable struct Book <: Model
    name::Field = Field(type="", descripton="", ...) 
    ...
end

@get "/book" function(book::Book)
    # handle book
end
```
An OpenAPI spec is generated automatically based on those, but this may require changes within Oxygen.jl, so it could be a rabbit hole while you have a solution already but nice to have.

**Reply by ndortega** [reply]
I like the ideas! I should be able to incorporate the Field auto-documentation. But the others I'm a bit wary of.

Authentication
- I'm not a security expert and I'd be wary of trying to write a one-size-fits-all module to do so. 
- FastApi provides some helper functions to help you set it up but doesn't do it for you ( I'd be comfortable trying this approach )

Sessions
- This is a bit more straightforward than the others and I could probably add this

database & templating 
- While there are some sweet libraries out there that can do both, I'd rather not tie myself to any one implementation. Mustache.jl is pretty cool, but other templating engines might show up in the future. Personally, I don't want to take away the decision from the developer because all projects differ in scope and size and there's no such thing as a "silver bullet".

**Reply by HiperMaximus** [reply]
It's not necessary that you tie yourself with some packages, but it'll be nice to have examples with different packages, for example with Octo.jl for database management.

**Reply by HiperMaximus** [reply]
Also it would be nice to have examples with some Graphql libraries and how to do OAuth2/OpenID

**Reply by jiachengzhang1** [reply]
I agree, examples and tutorials would be awesome. I'd like to contribute if you need help.

Another idea to make Oxygen "unique" comparing to other Julia web frameworks, is to add some kind of middleware chaining like [express.js](https://expressjs.com), I found it useful sometime. [Mux.jl](https://github.com/JuliaWeb/Mux.jl) has `stack` but it's not the cleanest. It could look like,

```julia
auth_handler = function(req, res) ... end
validation_handler = function(req, res) ... end
find_book = function(req, res) ... end

@get "/book" [auth_handler, validation_handler, find_book]
```

If you look for OAuth2, I have been working on [Umbrella.jl](https://github.com/jiachengzhang1/Umbrella.jl) to aid integrating OAuth2 servers like Google, GitHub and Facebook, feel free to steal the example there 😃

**Reply by ndortega** [reply]
@jiachengzhang1 
I'll gladly take you up on that offer! Creating examples and documentation takes a lot of effort and would help me out considerably. By the way, Umbrella.jl looks awesome and I'll gladly steal your example for the docs. I'm currently working on the documentation, so I'll make sure to add a section just for this (feel free to open a PR to change it up)

Regarding the second point about adding chained middleware: we won't have to implement that. The [latest changes](https://juliaweb.github.io/HTTP.jl/stable/#%22Handlers%22-framework-overhaul) in the HTTP.jl 1.0.0 release did some major refactoring and follows that pattern but still draws the distinction between handlers vs middleware. This will be available once I upgrade the framework, which will require some work from my end since a bunch of things have changed since then.

**Reply by HiperMaximus** [reply]
Really excited to see those changes!
